Add Declarative Web Push flags and JSON parser

beidson opened the pull request:
#### 15e5e670c08d732956d975c9829525d0b0f0f31e
<pre>
Add Declarative Web Push flags and JSON parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=261588">https://bugs.webkit.org/show_bug.cgi?id=261588</a>

Reviewed by Chris Dumez.

This adds:
- A Declarative Web Push runtime feature flag
- A Declarative Web Push compile time ENABLE flag
- The Notification payload JSON parser
- The beginnings of webpushd understanding a JSON message
- Testing the JSON parser by messaging JSON to webpushd

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationJSONParser.cpp: Added.
(WebCore::dirKey):
(WebCore::langKey):
(WebCore::bodyKey):
(WebCore::tagKey):
(WebCore::iconKey):
(WebCore::dataKey):
(WebCore::silentKey):
(WebCore::defaultActionURLKey):
(WebCore::titleKey):
(WebCore::appBadgeKey):
(WebCore::optionsKey):
(WebCore::parseAppBadgeString):
(WebCore::NotificationJSONParser::parseNotificationPayload):
(WebCore::NotificationJSONParser::parseNotificationOptions):
* Source/WebCore/Modules/notifications/NotificationJSONParser.h: Added.
* Source/WebCore/Modules/notifications/NotificationPayload.cpp: Added.
(WebCore::NotificationPayload::parseJSON):
* Source/WebCore/Modules/notifications/NotificationPayload.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/PushMessageForTesting.h:
(WebKit::WebPushD::PushMessageForTesting::encode const):
(WebKit::WebPushD::PushMessageForTesting::decode):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::injectPushMessageForTesting):
* Source/WebKit/webpushd/WebPushDaemon.h:
(WebPushD::WebPushDaemon::setMachServiceName):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::getPendingPushMessages):
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::sendPushMessage):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
(pushMessageFromArguments):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/268032@main">https://commits.webkit.org/268032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/996e10d1821b74abd46447fbaa125e13c2985320

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18437 "90 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20287 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17264 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18633 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18927 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21166 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16820 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15986 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16990 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21191 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17728 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14904 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21798 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16645 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5333 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4387 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21011 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23038 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17403 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5181 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/notifications/NotificationJSONParser.h:
```diff
@@ -0,0 +1,46 @@
+
+namespace WebCore {
+
+class NotificationJSONParser {
```
Seems we could just use a namespace and free functions.

Doesn't this class get a default constructor and stuff? :)

cdumez commented on Source/WebCore/Modules/notifications/NotificationPayload.h:
```diff
@@ -0,0 +1,50 @@
+#include <wtf/URL.h>
+#include <wtf/text/WTFString.h>
+
+OBJC_CLASS NSDictionary;
```
Seems unnecessary.

cdumez commented on Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
```diff
@@ -20,7 +20,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-class WTF::URL {
+[WebKitPlatform] class WTF::URL {
```
Out of curiosity, what does this do? First time I see it.

beidson replied:
It puts the symbol in the library libWebKitPlatform.a, which is statically linkable by DRT/WKTR/TWKA for testing.

cdumez commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -264,22 +266,37 @@ static void tryCloseRequestConnection(xpc_object_t request)
-        connection.broadcastDebugMessage("Attempting to inject a push message from an unentitled process"_s);
-        replySender(false);
+        auto error = "Attempting to inject a push message from an unentitled process"_s;
+        connection.broadcastDebugMessage(error);
```
This is a bit inefficient is constructs a String from the `error` ASCIILiteral twice, here and on the line below.
It would be better to use this above:
```
String error = "Attempting to inject a push message from an unentitled process"_s;
```

cdumez commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -264,22 +266,37 @@ static void tryCloseRequestConnection(xpc_object_t request)
-        connection.broadcastDebugMessage("Attempting to inject an invalid push message"_s);
-        replySender(false);
+        auto error = "Attempting to inject an invalid push message"_s;
+        connection.broadcastDebugMessage(error);
```
ditto.

cdumez reviewed (approved):
LGTM with nits.

beidson committed:
Add Declarative Web Push flags and JSON parser
https://bugs.webkit.org/show_bug.cgi?id=261588

Reviewed by Chris Dumez.

This adds:
- A Declarative Web Push runtime feature flag
- A Declarative Web Push compile time ENABLE flag
- The Notification payload JSON parser
- The beginnings of webpushd understanding a JSON message
- Testing the JSON parser by messaging JSON to webpushd

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationJSONParser.cpp: Added.
(WebCore::dirKey):
(WebCore::langKey):
(WebCore::bodyKey):
(WebCore::tagKey):
(WebCore::iconKey):
(WebCore::dataKey):
(WebCore::silentKey):
(WebCore::defaultActionURLKey):
(WebCore::titleKey):
(WebCore::appBadgeKey):
(WebCore::optionsKey):
(WebCore::parseAppBadgeString):
(WebCore::NotificationJSONParser::parseNotificationPayload):
(WebCore::NotificationJSONParser::parseNotificationOptions):
* Source/WebCore/Modules/notifications/NotificationJSONParser.h: Added.
* Source/WebCore/Modules/notifications/NotificationPayload.cpp: Added.
(WebCore::NotificationPayload::parseJSON):
* Source/WebCore/Modules/notifications/NotificationPayload.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/PushMessageForTesting.h:
(WebKit::WebPushD::PushMessageForTesting::encode const):
(WebKit::WebPushD::PushMessageForTesting::decode):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::injectPushMessageForTesting):
* Source/WebKit/webpushd/WebPushDaemon.h:
(WebPushD::WebPushDaemon::setMachServiceName):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::getPendingPushMessages):
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::sendPushMessage):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
(pushMessageFromArguments):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: https://commits.webkit.org/268032@main
